File: src/main/java/dragonbook/symbols/Array.java
package dragonbook.symbols;

import dragonbook.lexer.Tag;

public class Array extends Type {
    public final Type of; // array *of* type
    public final int size; // number of elements

    public Array(int size, Type of) {
        super("[]", Tag.INDEX, size * of.width);
        this.size = size;
        this.of = of;
    }

    public String toString() {
        return of + "[" + size + "]";
    }
}

File: src/main/kotlin/dragonbook/symbols/Array.kt
package dragonbook.symbols

import dragonbook.lexer.Tag

data class Array(
    val of: Type, // array *of* type
    val size: Int // number of elements
) : Type {

    override val tag: Int get() = Tag.INDEX
    override val lexeme: String get() = "[]"
    override val width: Int get() = size * of.width
    override fun toString(): String = "$of[$size]"

}